### ISSUE TITLE
Documentation Branch

### ISSUE THREAD

**Issue by etobella** [OCA/OpenUpgrade#3702]
@sbidoul @bguillot @pedrobaeza @legalsylvain 

We were reviewing the documentation and it is confusing to have a documentation folder for each version, 

Wouldn't have more sense to create a new repo or branch for OpenUpgrade documentation only?

**Comment by pedrobaeza** [comment]
Not really, as the module coverage should be update with each PR integrating the migration scripts. It would be a lot of work to synchronize that with 2 repos.

**Comment by pedrobaeza** [comment]
The rule is:

- Main explanations from latest branch.
- Module coverage taken from each branch.

**Comment by legalsylvain** [comment]
@pedrobaeza, @StefanRijnhart : do you how the branch is selected to generate the website ? https://oca.github.io/OpenUpgrade/
(I don't have access to the "settings / page" part of this repo).
Thanks.

**Comment by StefanRijnhart** [comment]
https://github.com/OCA/OpenUpgrade/settings/pages
![image](https://user-images.githubusercontent.com/1033124/214608477-566f6f80-7bcd-4de6-a53e-dc3deef64584.png)

**Comment by etobella** [comment]
@pedrobaeza We were talking about how to do things, not the coverage of each module. 

There is a documentation of how to do a migration on https://github.com/OCA/OpenUpgrade/blob/15.0/docsource

And this information might not be ported from version to version. It mitght be cleaner to add this on a separate branch / repo. 

Obviously, the current version configuration must be on the expected branch

**Comment by StefanRijnhart** [comment]
I imagine we can switch the documentation to 16.0 soon, once https://github.com/OCA/OpenUpgrade/pull/3680 is merged.

**Comment by legalsylvain** [comment]
hi. 

I don't understand your points @pedrobaeza and @StefanRijnhart.

documentation and openupgrade scripts have different lifecycles.
- openupgrade script (and framework) are related to a version, so having a branch per version is perfect.
- documentation is for all the version, so having documentation in each branch doesn't makes sense.

So :
- Each time a new release is out and became the default one, it makes obsolete all the previous documentation PR.
A simple exemple is regarding [the use case "field renaming"](https://github.com/OCA/OpenUpgrade/pull/3493) PR.
- The PR has been done against the 15.0 branch (July 2022). 
- The branch 16.0 has been created in October 2022.
- The PR has been merged after the creation of the 16.0

As a result as soon as the documentation branch is 15.0, the text is available on the doc [here](https://oca.github.io/OpenUpgrade/use_cases/field_renaming.html). But once the documentation branch will be the 16.0, the text will disappear. And we have to make a new PR to cherry pick the documentation. It requires extra useless work.

in addition, currently, "both" CIs are executed which does not make sense. 
- If we do a PR on the framework, there is no point in playing the documentation CI
- If we play a PR on the documentation, running the Openupgrade tests is useless.

> I imagine we can switch the documentation to 16.0 soon, once https://github.com/OCA/OpenUpgrade/pull/3680 is merged.

I don't get it. What prevent to switch to 16.0 ? 


Well, as @etobella proposed, creating a new branch named "documentation" and move all the documentation stuff (docsource + documentation CI) should work.
What exactly are your concerns with this change? 

Thanks.

**Comment by pedrobaeza** [comment]
You are forgetting this kind of file: https://github.com/OCA/OpenUpgrade/blob/15.0/docsource/modules140-150.rst, that is updated with each PR. There hasn't been the switch because the file `modules150-160.rst` hasn't been prepared yet, and the Holger's PR will contain it.

**Comment by legalsylvain** [comment]
Well. The proposal is to create a documentation branch and to move all the documentation stuff *except* the coverage file that should remain in each branch

**Comment by legalsylvain** [comment]
> There hasn't been the switch because the file modules150-160.rst hasn't been prepared yet, and the Holger's PR will contain it.

- The 15.0 16.0 file existe in V16 : https://github.com/OCA/OpenUpgrade/blob/16.0/docsource/modules150-160.rst

**Comment by pedrobaeza** [comment]
Well, it's more than that, as the index should be updated, but when done, it can be switched indeed.

About moving some stuff to other branch, the documentation is always on the latest branch, so having the documentation across 2 branches is far from ideal IMO.

**Comment by StefanRijnhart** [comment]
I'm suggesting that we switch branches to 16.0 only after base is merged because ideally, the branch that is used for the documentation should be the branch where the focus is on, and most of the changes. Otherwise, the built documentation does not reflect the progress on the (other) most active branch.
I don't see how splitting off the documentation solves a lot of problems, and I kind of like that the documentation is distributed with the software, as is common practice.

**Comment by legalsylvain** [comment]
> I'm suggesting that we switch branches to 16.0 only after base is merged because ideally, the branch that is used for the documentation should be the branch where the focus is on, and most of the changes. Otherwise, the built documentation does not reflect the progress on the (other) most active branch.

OK. i see ! Sorry I didn't understand your point. Well for that topic, it exists a "schedule" syntax in the github workflows. adding such "schedule" in a new dedicated workflow file in the default branch make the update of the website automatic. So we can define that each hour / day / ... the documentation is build and deployed on the website. No more need to target "the most active branch" as we do for the time being.

 >  I kind of like that the documentation is distributed with the software, as is common practice.

if documentation depends on each branch, It's normal to set it in each branch, yes. But in our case : 
- the objective of the documentation is to present how Openupgrade work for ALL the odoo versions. So having a dedicated branch makes sense.
- In fact, this doc is not only about https://github.com/OCA/OpenUpgrade project but also https://github.com/OCA/openupgradelib (See [website page](https://oca.github.io/OpenUpgrade/API.html#module-openupgradelib.openupgrade))

> I don't see how splitting off the documentation solves a lot of problems

Well, as said previously. 
1) Regarding the RFQ, in the current situation : 
- either I make changes against version 15, and once merged they will be visible on the website, but will disappear when we switch to v16
- or I make changes against version 16, and once merged, they won't be visible on the website  until the switch is made
2) The contribution made in https://github.com/OCA/OpenUpgrade/pull/3493 and merged will be lost when the switch to V16 will be done. We have to make a new useless PR to cherry pick the changes.

### to move on : 

I made a PoC of what I propose. Maybe it will be more clear for all.

- The PoC is available here : https://github.com/grap/openupgrade-doc-test/tree/documentation
- Documentation is here : https://grap.github.io/openupgrade-doc-test/
- My proposal is to create a new branch named "documentation" in github.com/OCA/Openupgrade that contains exactly the same things
- As you can see, schedule is set to refresh the website regularly [in that worflow, here](https://github.com/grap/openupgrade-doc-test/blob/master/.github/workflows/documentation-commit-schedule.yml). As a result, build are run regularly (see github actions logs [here](https://github.com/grap/openupgrade-doc-test/actions?query=event%3Aschedule)) and if there is a change in ANY coverage file in ANY branch (and in openupgradelib), it will be displayed in 24 hours.
- On that documentation branch, no need to run the validity tests of scripts present in the "test.yml" file. So documentation CI is lighter : https://github.com/grap/openupgrade-doc-test/tree/documentation/.github/workflows

Then once done, in a second step : 

we could remove from the regular branches (15.0, 16.0, etc...) : 
- .github/workflows/documentation-commit.yml file
- .github/workflows/documentation.yml file
- all the docs folder
- all the content of the docsource folder EXCEPT the file docsource/modulesXXX-YYY.rst 

we should add in the default branch the [schedule workflow file](https://github.com/grap/openupgrade-doc-test/blob/master/.github/workflows/documentation-commit-schedule.yml). 

Could you take a look on that PoC, and let me know if if you see any problems ?

Thanks !

**Comment by legalsylvain** [comment]
@StefanRijnhart : did you took a look on my last answer ? i think the PoC I did solves your question regarding the regular refresh of the documentation. 
Thanks !

**Comment by pedrobaeza** [comment]
Already talked directly. Final decision in hands of board.

**Comment by bguillot** [comment]
Hello, 

After discussion with the board, we are in favor of Sylvain's proposal.
The idea of having the documentation in one branch seems good.

@legalsylvain, you can start the documentation branch and we will ensure your POC fits the needs.

Thanks
